xray/segment.go

```diff
@@ -164,15 +158,44 @@ func (seg *Segment) RemoveSubsegment(remove *Segment) bool {
 	return false
 }
 
+func (seg *Segment) handleContextDone() {
+	seg.Lock()
+	defer seg.Unlock()
+
+	seg.ContextDone = true
+	if !seg.InProgress && !seg.Emitted {
+		seg.flush()
+	}
+}
+
+func (seg *Segment) flush() {
+	if (seg.openSegments == 0 && seg.EndTime > 0) || seg.ContextDone {
+		if seg.parent == nil {
+			seg.Emitted = true
+			emit(seg)
+		} else {
+			seg.parent.safeFlush()
+		}
+	}
+}
+
+func (seg *Segment) safeFlush() {
```

Comment:
`safe` seems like the wrong name here. This is `flush(decrement: true)` so... `flushWithDecrement`? ew that's a gross name to...

Perhaps it's difficult to name because it's not factored along a meaningful axis?